beacon_node/http_api/src/publish_attestations.rs

```diff
@@ -0,0 +1,292 @@
+//! validator is already known in this epoch.
+//!
+//! There's little value for the network in validating a second
+//! attestation for another validator since it is either:
+//!
+//! 1. A duplicate.
+//! 2. Slashable.
+//! 3. Invalid.
+//!
+//! We are likely to get duplicates in the case where a VC is using
+//! fallback BNs. If the first BN actually publishes some/all of a
+//! batch of attestations but fails to respond in a timely fashion,
+//! the VC is likely to try publishing the attestations on another
+//! BN. That second BN may have already seen the attestations from
+//! the first BN and therefore indicate that the attestations are
+//! "already seen". An attestation that has already been seen has
+//! been published on the network so there's no actual error from
+//! the perspective of the user.
+//!
+//! It's better to prevent slashable attestations from ever
+//! appearing on the network than trying to slash validators,
+//! especially those validators connected to the local API.
+//!
+//! There might be *some* value in determining that this attestation
+//! is invalid, but since a valid attestation already it exists it
+//! appears that this validator is capable of producing valid
+//! attestations and there's no immediate cause for concern.
+use crate::task_spawner::{Priority, TaskSpawner};
+use beacon_chain::{
+    validator_monitor::timestamp_now, AttestationError, BeaconChain, BeaconChainError,
+    BeaconChainTypes,
+};
+use beacon_processor::work_reprocessing_queue::{QueuedUnaggregate, ReprocessQueueMessage};
+use eth2::types::Failure;
+use lighthouse_network::PubsubMessage;
+use network::NetworkMessage;
+use slog::{debug, error, Logger};
+use std::sync::Arc;
+use std::time::Duration;
+use tokio::sync::{
+    mpsc::{Sender, UnboundedSender},
+    oneshot,
+};
+use types::Attestation;
+
+#[derive(Debug)]
+enum Error {
+    Validation(AttestationError),
+    Publication,
+    ForkChoice(BeaconChainError),
+    AggregationPool(AttestationError),
+    ReprocessDisabled,
+    ReprocessFull,
+    ReprocessTimeout,
+}
+
+enum PublishAttestationResult {
+    Success,
+    AlreadyKnown,
+    Reprocessing(oneshot::Receiver<Result<(), Error>>),
+    Failure(Error),
+}
+
+fn verify_and_publish_attestation<T: BeaconChainTypes>(
+    chain: &Arc<BeaconChain<T>>,
+    attestation: &Attestation<T::EthSpec>,
+    seen_timestamp: Duration,
+    network_tx: &UnboundedSender<NetworkMessage<T::EthSpec>>,
+) -> Result<(), Error> {
+    let attestation = chain
+        .verify_unaggregated_attestation_for_gossip(attestation, None)
+        .map_err(Error::Validation)?;
+
+    // Publish.
+    network_tx
+        .send(NetworkMessage::Publish {
+            messages: vec![PubsubMessage::Attestation(Box::new((
+                attestation.subnet_id(),
+                attestation.attestation().clone(),
+            )))],
+        })
+        .map_err(|_| Error::Publication)?;
+
+    // Notify the validator monitor.
+    chain
```

Comment:
Small order change here which seems intentional and reasonable, but thought I'd confirm anyway - we previously notify the monitor first - so it gets tracked regardless of whether publish is successful, was the original behaviour a bug thats fixed here?

Reply:
Yeah I think this order makes more sense:

- Attestations only get tracked if we published them successfully, and
- We prioritise publishing ASAP.